cocotb/handle.py

```diff
@@ -625,7 +625,7 @@ def setimmediatevalue(self, value):
             self._log.critical("Unsupported type for value assignment: %s (%s)", type(value), repr(value))
             raise TypeError("Unable to set simulator value with type %s" % (type(value)))
 
-        simulator.set_signal_val_str(self._handle, value.binstr)
+        simulator.set_signal_val_binstr(self._handle, value.binstr.encode())
```

Comment:
There is a call to `simulator.set_signal_val_str` below that also needs the string to bytes conversion.

Reply:
I'd argue that for `set_signal_val_str` by design, and that the user should perform the conversion. As far as I can tell, systemverilog has no concept of encoding, and just copies across raw bytes.

Reply:
It's more of a convenience thing. People expect sending `"TEST"` off to just work. We could add a check in here to try `value.encode('ascii')` if given a `str` object, pass through bytes, or complain otherwise.

Reply:
I understand where you're coming from, but python 3 aimed to draw the line at that type of behavior.

Reply:
And passing `b"test"` isn't much harder

Reply:
It will break existing correct sims. Also, we need to reconsider what gets return by `get_signal_value_str` and `get_signal_value_binstr`. This is more of a change than a fix, and all the tests that I intend to fix will still work despite this change. So, I will remove this commit leaving it parsed with `s` and open a separate PR.

Reply:
I don't necessarily agree that such sims are _correct_, but if they're already executing I'm with you that we don't want to change the behavior in this PR.